rfcs/rfc-001-smart-contracts-engine.md

```diff
@@ -37,154 +37,175 @@ version 0.1
-
-## Loader
-The loader is our first smart contract. The job of this contract is to load the actual program with its own instance data.  The loader will verify the bytecode and that the object implements the expected entry points.
-
-Since there is only one RX segment, the context for the contract instance is passed into each entry point as well as the event data for that entry point.
-
-A client will create a transaction to create a new loader instance:
-
-`Solana_NewLoader(Loader Instance PubKey, proof of key ownership, space I need for my elf)`
-
-A client will then do a bunch of transactions to load its elf into the loader instance they created:
-
-`Loader_UploadElf(Loader Instance PubKey, proof of key ownership, pos start, pos end, data)`
-
-At this point the client can create a new instance of the module with its own instance address:
-
-`Loader_NewInstance(Loader Instance PubKey, proof of key ownership, Instance PubKey, proof of key ownership)`
-
-Once the instance has been created, the client may need to upload more user data to solana to configure this instance:
-
-`Instance_UploadModuleData(Instance PubKey, proof of key ownership, pos start, pos end, data)`
-
-Now clients can `start` the instance:
-
-`Instance_Start(Instance PubKey, proof of key ownership)`
-
 ## Runtime
 
-Our goal with the runtime is to have a general purpose execution environment that is highly parallelizable and doesn't require dynamic resource management. We want to execute as many contracts as we can in parallel, and have them pass or fail without a destructive state change.
-
-### State and Entry Point
-
-State is addressed by an account which is at the moment simply the PubKey.  Our goal is to eliminate dynamic memory allocation in the smart contract itself, so the contract is a function that takes a mapping of [(PubKey,State)] and returns [(PubKey, State')].  The output of keys is a subset of the input.  Three basic kinds of state exist:
+The goal with the runtime is to have a general purpose execution environment that is highly parallelizeable and doesn't require dynamic resource management. The goal is to execute as many contracts as possible in parallel, and have them pass or fail without a destructive state change.
 
-* Instance State
-* Participant State
-* Caller State
 
-There isn't any difference in how each is implemented, but conceptually Participant State is memory that is allocated for each participant in the contract.  Instance State is memory that is allocated for the contract itself, and Caller State is memory that the transactions caller has allocated.
+### State
 
+State is addressed by an account which is at the moment simply the PubKey.  Our goal is to eliminate dynamic memory allocation in the smart contract itself, so the contract is a function that takes a mapping of [(PubKey,State)] and returns [(PubKey, State')].
```

Comment:
It's `Pubkey` with a lowercase "k". The goal is not to eliminate dynamic memory allocation, but instead to offer dynamic memory for rent so that the runtime doesn't need to put artificial constraints on the contract to limit memory usage.

A contract isn't the function you described, which I'd call a state transition function. Instead, the runtime interacts with a contract through its state transition function, which the runtime will invoke many times to form what's traditionally been called a smart contract.

The type of the state transition function is conceptually what you've described, but that wouldn't compile unless `State` is a type alias for some opaque type like `Vec<u8>`.